src/deluge/model/song/song.cpp

```diff
@@ -2228,6 +2228,17 @@ Error Song::readClipsFromFile(Deserializer& reader, ClipArray* clipArray) {
 				delugeDealloc(memory);
 				return error;
 			}
+			// find and link to its group if needed
+			if (newClip->isInGroup()) {
+				auto c = clipArray->getNumElements();
+				for (c; c >= 0; c--) {
+					auto prevClip = clipArray->getClipAtIndex(c);
```

Comment:
I feel like I must be misreading this, but isn't this iterating backwards from N, instead of N-1? Ie. isn't the first access out of bounds?

I would have expected

```
for (auto c = clipArray->getNumElements() - 1; c >= 0; c--) {
```

Reply:
I think you might have been reviewing a specific commit and not the whole pr - you're correct that this was a bug but I fixed it already haha 


Same for moving the insert function into the cpp file instead of the header